diff -aruN php-5.6.40.orig/ext/pdo_firebird/firebird_driver.c php-5.6.40/ext/pdo_firebird/firebird_driver.c
--- php-5.6.40.orig/ext/pdo_firebird/firebird_driver.c	2021-07-15 05:38:27.640549401 -0400
+++ php-5.6.40/ext/pdo_firebird/firebird_driver.c	2021-07-15 05:40:37.866105581 -0400
@@ -530,15 +530,16 @@
 	return 0;
 }
 /* }}} */
+#define INFO_BUF_LEN 512
 
 /* callback to used to report database server info */
 static void firebird_info_cb(void *arg, char const *s) /* {{{ */
 {
 	if (arg) {
 		if (*(char*)arg) { /* second call */
-			strcat(arg, " ");
+			strlcat(arg, " ", INFO_BUF_LEN);
 		}
-		strcat(arg, s);
+		strlcat(arg, s, INFO_BUF_LEN);
 	}
 }
 /* }}} */
@@ -549,7 +550,7 @@
 	pdo_firebird_db_handle *H = (pdo_firebird_db_handle *)dbh->driver_data;
 
 	switch (attr) {
-		char tmp[512];
+		char tmp[INFO_BUF_LEN];
 		
 		case PDO_ATTR_AUTOCOMMIT:
 			ZVAL_LONG(val,dbh->auto_commit);
diff -aruN php-5.6.40.orig/ext/pdo_firebird/tests/bug_76448.bug php-5.6.40/ext/pdo_firebird/tests/bug_76448.bug
--- php-5.6.40.orig/ext/pdo_firebird/tests/bug_76448.bug	1969-12-31 19:00:00.000000000 -0500
+++ php-5.6.40/ext/pdo_firebird/tests/bug_76448.bug	2021-07-15 05:42:21.343136719 -0400
@@ -0,0 +1,23 @@
+--TEST--
+Bug #76448 (Stack buffer overflow in firebird_info_cb)
+--SKIPIF--
+<?php
+if (!extension_loaded('pdo_firebird')) die("skip podo_firebird extension not available");
+if (!extension_loaded('sockets')) die("skip sockets extension not available");
+?>
+--FILE--
+<?php
+require_once "payload_server.inc";
+
+$address = run_server(__DIR__ . "/bug_76448.data");
+
+// no need to change the credentials; we're running against a falke server
+$dsn = "firebird:dbname=inet://$address/test";
+$username = 'SYSDBA';
+$password = 'masterkey';
+
+$dbh = new PDO($dsn, $username, $password, [PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION]);
+var_dump($dbh->getAttribute(PDO::ATTR_SERVER_INFO));
+?>
+--EXPECT--
+bool(false)
\ No newline at end of file
