diff --git a/ext/dom/php_dom.c b/ext/dom/php_dom.c
index 4a490b2..766f776 100644
--- a/ext/dom/php_dom.c
+++ b/ext/dom/php_dom.c
@@ -1534,6 +1534,14 @@ void dom_normalize (xmlNodePtr nodep TSRMLS_DC)
 						break;
 					}
 				}
+				strContent = xmlNodeGetContent(child);
+				if (*strContent == '\0') {
+					nextp = child->next;
+					xmlUnlinkNode(child);
+					php_libxml_node_free_resource(child);
+					child = nextp;
+					continue;
+				}
 				break;
 			case XML_ELEMENT_NODE:
 				dom_normalize (child TSRMLS_CC);
