diff --git a/ext/date/php_date.c b/ext/date/php_date.c
index 47f78a2..edb896b 100644
--- a/ext/date/php_date.c
+++ b/ext/date/php_date.c
@@ -630,6 +630,7 @@ static zend_object_value date_object_clone_interval(zval *this_ptr TSRMLS_DC);
 static zend_object_value date_object_clone_period(zval *this_ptr TSRMLS_DC);
 
 static int date_object_compare_date(zval *d1, zval *d2 TSRMLS_DC);
+static int date_object_compare_timezone(zval *tz1, zval *tz2 TSRMLS_DC);
 static HashTable *date_object_get_gc(zval *object, zval ***table, int *n TSRMLS_DC);
 static HashTable *date_object_get_properties(zval *object TSRMLS_DC);
 static HashTable *date_object_get_gc_interval(zval *object, zval ***table, int *n TSRMLS_DC);
@@ -2058,6 +2059,7 @@ static void date_register_classes(TSRMLS_D)
 	date_ce_timezone = zend_register_internal_class_ex(&ce_timezone, NULL, NULL TSRMLS_CC);
 	memcpy(&date_object_handlers_timezone, zend_get_std_object_handlers(), sizeof(zend_object_handlers));
 	date_object_handlers_timezone.clone_obj = date_object_clone_timezone;
+	date_object_handlers_timezone.compare_objects = date_object_compare_timezone;
 	date_object_handlers_timezone.get_properties = date_object_get_properties_timezone;
 	date_object_handlers_timezone.get_gc = date_object_get_gc_timezone;
 
@@ -2191,6 +2193,40 @@ static int date_object_compare_date(zval *d1, zval *d2 TSRMLS_DC)
 	return (o1->time->sse == o2->time->sse) ? 0 : ((o1->time->sse < o2->time->sse) ? -1 : 1);
 }
 
+static int date_object_compare_timezone(zval *tz1, zval *tz2 TSRMLS_DC) /* {{{ */
+{
+	php_timezone_obj *o1, *o2;
+
+	if (Z_TYPE_P(tz1) == IS_OBJECT && Z_TYPE_P(tz2) == IS_OBJECT &&
+		instanceof_function(Z_OBJCE_P(tz1), date_ce_timezone TSRMLS_CC) &&
+		instanceof_function(Z_OBJCE_P(tz2), date_ce_timezone TSRMLS_CC)) {
+		php_timezone_obj *o1 = zend_object_store_get_object(tz1 TSRMLS_CC);
+		php_timezone_obj *o2 = zend_object_store_get_object(tz2 TSRMLS_CC);
+		
+		if (!o1->initialized || !o2->initialized) {
+			php_error_docref(NULL, E_WARNING, "Trying to compare an incomplete DateTimeZone object");
+			return 1;
+		}
+
+		if (o1->type != o2->type) {
+			php_error_docref(NULL, E_WARNING, "Trying to compare different kinds of DateTimeZone objects");
+			return 1;
+		}		
+		
+		switch (o1->type) {
+			case TIMELIB_ZONETYPE_OFFSET:
+				return o1->tzi.utc_offset == o2->tzi.utc_offset ? 0 : 1;
+			case TIMELIB_ZONETYPE_ABBR:
+				return strcmp(o1->tzi.z.abbr, o2->tzi.z.abbr) ? 1 : 0;
+			case TIMELIB_ZONETYPE_ID:
+				return strcmp(o1->tzi.tz->name, o2->tzi.tz->name) ? 1 : 0;
+			EMPTY_SWITCH_DEFAULT_CASE();
+		}
+	}
+	
+	return 1;
+} /* }}} */
+
 static HashTable *date_object_get_gc(zval *object, zval ***table, int *n TSRMLS_DC)
 {
 	*table = NULL;
