diff --git a/ext/standard/file.c b/ext/standard/file.c
index 067f961..f5bb6ee 100644
--- a/ext/standard/file.c
+++ b/ext/standard/file.c
@@ -1624,7 +1624,7 @@ PHP_FUNCTION(copy)
 		return;
 	}
 
-	if (php_check_open_basedir(source TSRMLS_CC)) {
+	if (php_stream_locate_url_wrapper(source, NULL, 0 TSRMLS_CC) == &php_plain_files_wrapper && php_check_open_basedir(source TSRMLS_CC)) {
 		RETURN_FALSE;
 	}
 
